File: python/foundry-local-whisper.py
"""
Whisper Voice Transcription with Foundry Local
Transcribes WAV audio files using the OpenAI Whisper model running locally.
Usage: python foundry-local-whisper.py [path-to-wav-file]
"""

import sys
import os
import glob
import openai
from foundry_local import FoundryLocalManager

model_alias = "whisper-medium"

# Default: transcribe all Zava sample WAV files
samples_dir = os.path.join(os.path.dirname(__file__), "..", "samples", "audio")

# If a specific file is provided as argument, transcribe only that file
if len(sys.argv) > 1:
    audio_files = [sys.argv[1]]
    for f in audio_files:
        if not os.path.exists(f):
            print(f"Audio file not found: {f}")
            print("Usage: python foundry-local-whisper.py [path-to-wav-file]")
            sys.exit(1)
else:
    audio_files = sorted(glob.glob(os.path.join(samples_dir, "zava-*.wav")))
    if not audio_files:
        print(f"No WAV files found in {samples_dir}")
        print("Run 'python samples/audio/generate_samples.py' first to generate them.")
        sys.exit(1)

# Step 1: Start the Foundry Local service
print("Starting Foundry Local service...")
manager = FoundryLocalManager()
manager.start_service()

# Step 2: Check if the model is already downloaded
cached = manager.list_cached_models()
catalog_info = manager.get_model_info(model_alias)
is_cached = any(m.id == catalog_info.id for m in cached) if catalog_info else False

if is_cached:
    print(f"Model already downloaded: {model_alias}")
else:
    print(f"Downloading model: {model_alias} (this may take several minutes)...")
    def on_progress(progress):
        bar_width = 30
        filled = int(progress / 100 * bar_width)
        bar = "\u2588" * filled + "\u2591" * (bar_width - filled)
        sys.stdout.write(f"\rDownloading: [{bar}] {progress:.1f}%")
        if progress >= 100:
            sys.stdout.write("\n")
        sys.stdout.flush()
    manager.download_model(model_alias, progress_callback=on_progress)
    print(f"Download complete: {model_alias}")

# Step 3: Load the model into memory
print(f"Loading model: {model_alias}...")
manager.load_model(model_alias)
model_id = manager.get_model_info(model_alias).id
print(f"Model loaded: {model_id}")

# Step 4: Create the OpenAI client pointing to the local service
client = openai.OpenAI(
    base_url=manager.endpoint,
    api_key=manager.api_key
)

# Step 5: Transcribe each audio file
for audio_path in audio_files:
    filename = os.path.basename(audio_path)
    print(f"\n{'=' * 60}")
    print(f"File: {filename}")
    print("=" * 60)

    with open(audio_path, "rb") as f:
        result = client.audio.transcriptions.create(
            model=model_id,
            file=f
        )

    print(result.text)

print(f"\nDone — transcribed {len(audio_files)} file(s).")

File: javascript/foundry-local-whisper.mjs
/**
 * Whisper Voice Transcription with Foundry Local
 * Transcribes WAV audio files using the OpenAI Whisper model running locally.
 * Usage: node foundry-local-whisper.mjs [path-to-wav-file]
 */

import { OpenAI } from "openai";
import { FoundryLocalManager } from "foundry-local-sdk";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const modelAlias = "whisper-medium";

// Default: transcribe all Zava sample WAV files
const samplesDir = path.join(__dirname, "..", "samples", "audio");

let audioFiles;
if (process.argv[2]) {
  // Specific file provided as argument
  const filePath = process.argv[2];
  if (!fs.existsSync(filePath)) {
    console.error(`Audio file not found: ${filePath}`);
    console.error("Usage: node foundry-local-whisper.mjs [path-to-wav-file]");
    process.exit(1);
  }
  audioFiles = [filePath];
} else {
  // Find all Zava WAV files
  if (!fs.existsSync(samplesDir)) {
    console.error(`Samples directory not found: ${samplesDir}`);
    console.error(
      "Run 'python samples/audio/generate_samples.py' first to generate them."
    );
    process.exit(1);
  }
  audioFiles = fs
    .readdirSync(samplesDir)
    .filter((f) => f.startsWith("zava-") && f.endsWith(".wav"))
    .sort()
    .map((f) => path.join(samplesDir, f));

  if (audioFiles.length === 0) {
    console.error(`No WAV files found in ${samplesDir}`);
    console.error(
      "Run 'python samples/audio/generate_samples.py' first to generate them."
    );
    process.exit(1);
  }
}

// Step 1: Bootstrap — starts the service, downloads, and loads the model
console.log(`Initializing Foundry Local with model: ${modelAlias}...`);
const manager = new FoundryLocalManager();
const modelInfo = await manager.init(modelAlias);
console.log(`Model ready: ${modelInfo.id}`);
console.log(`Endpoint: ${manager.endpoint}`);

// Step 2: Create the OpenAI client pointing to the local service
const client = new OpenAI({
  baseURL: manager.endpoint,
  apiKey: manager.apiKey,
});

// Step 3: Transcribe each audio file
for (const audioPath of audioFiles) {
  const filename = path.basename(audioPath);
  console.log(`\n${"=".repeat(60)}`);
  console.log(`File: ${filename}`);
  console.log("=".repeat(60));

  const result = await client.audio.transcriptions.create({
    model: modelInfo.id,
    file: fs.createReadStream(audioPath),
  });

  console.log(result.text);
}

console.log(`\nDone — transcribed ${audioFiles.length} file(s).`);
